_posts/http-client-factory-in-integration-testing.md

```diff
@@ -0,0 +1,160 @@
+---
+layout: post
+header_image: /img/posts/http-client-factory-in-integration-testing/header.jpg
+title: "HTTP client and HTTP client factory in integration testing"
+enable_mermaid: true
+author: Mauro Servienti
+synopsis: "Integration testing can be challenging. If we're also dealing with HTTP, it can be a nightmare. HTTP client factory in combination with the Microsoft.AspNetCore.Mvc.Testing package can come to the rescue and save our day."
+tags:
+- integration-testing
+---
+
+A few months ago in ["You wanna test HTTP, right?"](https://milestone.topics.it/2021/04/28/you-wanna-test-http.html), I presented how to self host a web application in a test using a self-contained approach that doesn't need external resources, like a running webserver. I didn't invent anything. I limited myself to showing how to use the `WebApplicationFactory<T>` shipped in the `Microsoft.AspNetCore.Mvc.Testing` package.
+
+## Are we done?
+
+I'm using the above-presented approach to design [ServiceComposer](https://github.com/ServiceComposer/ServiceComposer.AspNetCore) integration tests. For more details on ServiceComposer and what it does, refer to the articles in the [ViewModel Composition category](https://milestone.topics.it/categories/view-model-composition).
+
+ServiceComposer integration tests are pretty complex; let me try to simplify the scenario as much as possible. Let's take into account the following class:
+
+```csharp
+class AClassThatDoesSomething
+{
+   public async Task DoSomething()
+   {	 
+      var url = "http://domain.tld/do/something";
+      var httpClient = new HttpClient();
+      var response = await httpClient.GetAsync(url);
+      //do something meaningful with the response
+   }
+}
+```
+
+How could we test something like that? One option is to provide the test with a proper web server dependency. Not easy and hard to maintain as the complexity of the business logic increases. Another option could be to make the HTTP client a dependency injected via DI. At a first look, that's problematic too; we cannot mock HttpClient, too many options. Another apparent troubling aspect of injecting HttpClient is that the client instance is usually customized with default headers, a base address, and many other settings by each class using it. It would be cumbersome to move the required configuration into the DI/IoC infrastructure.
```

Comment:
```suggestion
How could we test something like that? One option is to provide the test with a proper web server dependency. Not easy, and hard to maintain as the complexity of the business logic increases. Another option could be to make the HTTP client a dependency injected via DI. At a first look, that's problematic too; we cannot mock HttpClient, there are too many options. Another apparent troubling aspect of injecting HttpClient is that the client instance is usually customized with default headers, a base address, and many other settings by each class using it. It would be cumbersome to move the required configuration into the DI/IoC infrastructure.
```